pandas_gbq/gbq.py

```diff
@@ -308,6 +312,104 @@ def run_query(self, query, **kwargs):
 
         return schema, result_rows
 
+    def export_table_to_gcs(
+        self,
+        dataset,
+        table,
+        timeout_in_seconds=None,
+        bucket=None,
+        blob=None,
+        zipped=True,
+    ):
+        """
+        export table to gcs.  returns tuple of (bucket, blob)
+        """
+        client = self.client
+        table_ref = client.dataset(dataset).table(table)
+        job_config = ExtractJobConfig()
+        job_config.compression = 'GZIP' if zipped else 'NONE'
+        bucket = bucket or '{}-temp'.format(client.project)
+        blob = blob or '{}/{}.csv'.format(dataset, table)
+        if zipped and not blob.endswith('.gz'):
+            blob += '.gz'
+        if not isinstance(bucket, str):
+            bucket = bucket.name
+        destination_uri = 'gs://{}/{}'.format(bucket, blob)
+        extract_job = client.extract_table(
+            table_ref, destination_uri, job_config=job_config)
+        wait_for_job(extract_job, timeout_in_seconds=timeout_in_seconds)
+        logger.info('Exported {}.{} -> {}'.format(
+            dataset,
+            table,
+            destination_uri)
+        )
+        return bucket, blob
+
+    def read_gbq_table(
+        self,
+        dataset,
+        table,
+        bucket,
+        timeout_in_seconds=600,
+    ):
+        """
+        reads an entire table from gbq into a dataframe
+        """
+        from google.cloud import storage
+
+        storage = storage.Client(self.project, self.credentials)
+        prefix = 'gbq-exports/{}/{}/'.format(dataset, table)
+        bucket = storage.get_bucket(bucket)
+
+        for old_blob in bucket.list_blobs(prefix=prefix):
+            old_blob.delete()
+            logger.info('Old Blob Deleted: {}'.format(old_blob.name))
+
+        self.export_table_to_gcs(
+            dataset=dataset,
+            table=table,
+            timeout_in_seconds=timeout_in_seconds,
+            bucket=bucket,
+            blob='{}*.csv.gz'.format(prefix),
+        )
+
+        frames = []
+
+        downloads = tqdm(list(bucket.list_blobs(prefix=prefix)), unit='file')
+        for blob in downloads:
+            downloads.set_description('Processing {}, {}MB'.format(
+                blob, blob.size / 2**20))
+            s = BytesIO(blob.download_as_string())
+            frames.append(pd.read_csv(s, compression='gzip'))
+            blob.delete()
+
+        return pd.concat(frames, ignore_index=True)
```

Comment:
This reads all the data into memory at once, right? In cases where the data size is too large to read into memory at once, it may make sense to run some of the above logic (everything before 376 I think), but not construct a data frame from the resulting files. (Clients may instead want to read individual files, or use Dask.) 

Would it be possible to refactor the code to make that possible?

Reply:
This would be interesting. Feel free to clone this PR and make this change. I agree this would be a good initial step for Dask integration.